src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/JsonFunctionCallUtilities.cs

```diff
@@ -0,0 +1,442 @@
+        }
+
+        return new FunctionCallContent(callId, functionName, arguments)
+        {
+            Exception = parsingException
+        };
+    }
+
+    /// <summary>Parses a JSON object into a <see cref="FunctionCallContent"/> with arguments encoded as <see cref="JsonElement"/>.</summary>
+    /// <param name="utf8Json">A UTF-8 encoded JSON object containing the parameters.</param>
+    /// <param name="callId">The function call ID.</param>
+    /// <param name="functionName">The function name.</param>
+    /// <returns>The parsed dictionary of objects encoded as <see cref="JsonElement"/>.</returns>
+    public static FunctionCallContent ParseFunctionCallContent(ReadOnlySpan<byte> utf8Json, string callId, string functionName)
+    {
+        _ = Throw.IfNull(callId);
+        _ = Throw.IfNull(functionName);
+
+        Dictionary<string, object?>? arguments = null;
+        Exception? parsingException = null;
+
+        try
+        {
+            arguments = JsonSerializer.Deserialize(utf8Json, FunctionCallUtilityContext.Default.DictionaryStringObject);
+        }
+        catch (JsonException ex)
+        {
+            parsingException = new InvalidOperationException($"Function call arguments contained invalid JSON: {Encoding.UTF8.GetString(utf8Json.ToArray())}", ex);
+        }
+
+        return new FunctionCallContent(callId, functionName, arguments)
+        {
+            Exception = parsingException
+        };
+    }
+
+    /// <summary>
+    /// Determines a JSON schema for the provided parameter metadata.
+    /// </summary>
+    /// <param name="parameterMetadata">The parameter metadata from which to infer the schema.</param>
+    /// <param name="functionMetadata">The containing function metadata.</param>
+    /// <param name="serializerOptions">The global <see cref="JsonSerializerOptions"/> governing serialization.</param>
+    /// <param name="inferenceOptions">The options controlling schema inference.</param>
+    /// <returns>A JSON schema document encoded as a <see cref="JsonElement"/>.</returns>
+    public static JsonElement InferParameterJsonSchema(
+        AIFunctionParameterMetadata parameterMetadata,
+        AIFunctionMetadata functionMetadata,
+        JsonSerializerOptions? serializerOptions,
+        JsonSchemaInferenceOptions? inferenceOptions = null)
+    {
+        _ = Throw.IfNull(parameterMetadata);
+        _ = Throw.IfNull(functionMetadata);
+
+        serializerOptions ??= functionMetadata.JsonSerializerOptions;
+
+        if (ReferenceEquals(serializerOptions, functionMetadata.JsonSerializerOptions) &&
+            parameterMetadata.Schema is JsonElement schema)
+        {
+            // If the resolved options matches that of the function metadata,
+            // we can just return the precomputed JSON schema value.
+            return schema;
+        }
+
+        if (serializerOptions is null)
+        {
+            return _trueJsonSchema;
+        }
```

Comment:
This seems like an odd policy. I don't provide a JSO so no inference takes place?

Reply:
Refactored so that it resolves to the source generated JSO.